docs/architecture.md

```diff
@@ -0,0 +1,545 @@
+
+### data flow
+
+https://mermaid.js.org/syntax/sequenceDiagram.html
+
+```mermaid
+---
+title: request flow
+---
+flowchart LR
+    user_defined_types
+    odata_request1[odata_request]
+    http_request1[http_request]
+    http_client
+    http_handler
+    http_request2[http_request]
+    odata_request2[odata_request]
+    data_store_request
+
+    user_defined_types --> odata_request1
+    odata_request1 --> http_request1
+    http_request1 --> http_client
+    http_client --> http_handler
+    http_handler --> http_request2
+    http_request2 --> odata_request2
+    odata_request2 --> data_store_request
+
+    style odata_request1 fill:#08FF08
+    style odata_request2 fill:#08FF08
+    style http_request1 fill:#FBF719
+    style http_request2 fill:#FBF719
+```
+
+```mermaid
+---
+title: response flow
+---
+flowchart LR
+    data_store_response
+    odata_response1[odata_response]
+    http_response1[http_response]
+    http_handler
+    http_client
+    http_response2[http_response]
+    odata_response2[odata_response]
+    user_defined_types
+
+    data_store_response --> odata_response1
+    odata_response1 --> http_response1
+    http_response1 --> http_handler
+    http_handler --> http_client
+    http_client --> http_response2
+    http_response2 --> odata_response2
+    odata_response2 --> user_defined_types
+
+    style odata_response1 fill:#08FF08
+    style odata_response2 fill:#08FF08
+    style http_response1 fill:#FBF719
+    style http_response2 fill:#FBF719
+```
+
+Notice that in each case, the request and response types are the same for both the client and the service. There are three side effects of this:
+1. The entire HTTP portion of the flow can be removed, meaning that **any** routing infrastructure can be leveraged while sharing the same common "odata_request" and "odata_response" types. This would allow ASP.NET services and azure functions to share the same basic scaffolding.
+2. Because the HTTP portion is not strictly necessary for processing a request, it means that odata services can be easily composed at this layer. If an odata_request is sent by a client using the same type that it is processed by the service, then it means that additional odata services can themselves be treated as "data stores". More on this [later](#a-common-abstraction-for-handling-a-request).
+3. These types can be shared by consumers of the odata libraries, but currently they are not. To see the impact of this sharing, let's dive into the mechanics of each translation of the data:
+
+#### user defined types to odata request
+
+```mermaid
+sequenceDiagram
+    participant user defined types
+    participant odata request AST
+
+    user defined types->>odata request AST: serialized to
+```
+
+#### odata request to http request
+
+```mermaid
+sequenceDiagram
+    participant odata request AST
+    participant odata request CST
```

Comment:
From @habbes : should we remove the CST and treat it as a "detail" or parsing the payloads?

Reply:
My gut reaction here is: it's almost "hard" to write the code **wrong** if there's both an AST and a CST. Because of this, I think it's really useful to have both concepts defined and implemented; it would be really useful to have those implementations to use as comparisons in test cases and such to make sure our transcribers, readers, and writers are implemented correctly. Those tests would be "complete" because they would be comparing the entire payload instead of just whatever assertions the test writer thought were meaningful at the time. This also could give us significantly better coverage on breaking changes. 

Let me think about this more.

Reply:
I'm going to keep the CST references in the doc because I think they will clarify later design decisions, but I've added a note that the CSTs won't necessarily be exposed to customers.